### PR TITLE
8.0 external file sync

### PR DESCRIPTION
**Pull request by Viggor**
After this discussion https://github.com/OCA/connector-interfaces/issues/5#issuecomment-72448656 i've create some module for unify the management of remote files. There is two module: 
- one add some fields to ir.attachment 
- the other manage import/export of ir.attachment. 

For the moment it's functionnal with FTP only but it will improved in the future with others protocol when the architecture will be validated.
None of this modules can parse a file, they only manage transmitting.
Thanks to @initOS @camptocamp and @akretion .

### CODE REVIEW

File: ir_attachment_metadata/attachment_view.xml [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+    <data>
+
+	    <record id="view_attachment_improved_form" model="ir.ui.view">
+	        <field name="model">ir.attachment</field>
+	    	<field name="inherit_id" ref="base.view_attachment_form" />
+	        <field name="arch" type="xml">
+		    <field name="url" position="after">
+			    <field name="internal_hash"/>
+			    <field name="external_hash"/>
+	    	    </field>
```
Comment (sebastienbeau) [inline ir_attachment_metadata/attachment_view.xml]: Fix indentation you have mixed tab and space here (visible in my vim)

File: external_file_location/location.py [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+#   GNU Affero General Public License for more details.
+#
+#   You should have received a copy of the GNU Affero General Public License
+#   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+###############################################################################
+
+from openerp import models, fields
+from knowledge.external_file_location.abstract_task import AbstractTask
+from knowledge.external_file_location.helper import itersubclasses
```
Comment (sebastienbeau) [inline external_file_location/location.py]: Wrong import use relative import from .abstract_task import AbstractTask

File: external_file_location/abstract_task.py [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+# -*- coding: utf-8 -*-
+
+from base64 import b64encode
+
+
+class AbstractTask(object):
+
+    def create_file(self, filename, data):
+        ir_attachment_id = self.env['ir.attachment'].create(
+                {
```
Comment (sebastienbeau) [inline external_file_location/abstract_task.py]: pep8

File: external_file_location/task.py [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+        tasks.run()
+
+    @api.one
+    def run(self):
+        for cls in itersubclasses(AbstractTask):
+            if cls._synchronize_type and \
+              cls._key + '_' + cls._synchronize_type == self.method:
+                method_class = cls
+        config = {
+                'host': self.location_id.address,
```
Comment (sebastienbeau) [inline external_file_location/task.py]: pep8

File: external_file_location/task.py [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+#   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+#   GNU Affero General Public License for more details.
+#
+#   You should have received a copy of the GNU Affero General Public License
+#   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+###############################################################################
+
+from openerp import models, fields, api
+from knowledge.external_file_location.helper import itersubclasses
```
Comment (sebastienbeau) [inline external_file_location/task.py]: wrong import

File: external_file_location/attachment_view.xml [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+            <field name="view_mode">kanban,tree,form</field>
+            <field name="view_id" eval="False"/>
+            <field name="domain">[('task_id', '!=', False)]</field>
+            <field name="search_view_id" ref="view_external_attachment_search"/>
+        </record>
+
+        <record id="ir_attachment_view1" model="ir.actions.act_window.view">
+            <field eval="20" name="sequence"/>
+            <field name="view_mode">kanban</field>
+            <field name="view_id" ref="mail.view_document_file_kanban"/>
```
Comment (sebastienbeau) [inline external_file_location/attachment_view.xml]: Are you sure that you want to use this view? If yes you should depend of the module "mail"

File: external_file_location/location_view.xml [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+
+        <record id="action_location" model="ir.actions.act_window">
+            <field name="name">Locations</field>
+            <field name="type">ir.actions.act_window</field>
+            <field name="res_model">external.file.location</field>
+            <field name="view_type">form</field>
+            <field name="view_id" eval="False"/>
+        </record>
+
+        <menuitem id="menu_ir_location"
```
Comment (sebastienbeau) [inline external_file_location/location_view.xml:62]: This child menu is defined before the parent menu. Installation fail
Change the order in **openerp**.py (data field) or group all the menu in th e same file (better option)

File: external_file_location/__openerp__.py [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+     'ir_attachment_metadata',
+ ],
+ 'data': [
+     'menu.xml',
+     'attachment_view.xml',
+     'location_view.xml',
+     'task_view.xml',
+     'cron.xml',
+ ],
+ 'installable': True,
```
Comment (bealdav) [inline external_file_location/__openerp__.py]: external_dependencies (/cy ?) key miss for ftputil.

File: external_file_location/__openerp__.py [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+ 'version': '0.0.1',
+ 'author': 'Akretion',
+ 'website': 'www.akretion.com',
+ 'license': 'AGPL-3',
+ 'category': 'Generic Modules',
+ 'description': """
+ File exchange system with multiple protocol (SFTP, FTP, Filestore)
+ """,
+ 'depends': [
+     'base',
```
Comment (bealdav) [inline external_file_location/__openerp__.py]: base could be removed

File: external_file_location/task_view.xml [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+
+        <record id="view_task_form" model="ir.ui.view">
+            <field name="model">external.file.task</field>
+            <field name="arch" type="xml">
+                <form string="Tasks" version="7.0">
+                    <sheet>
+                        <field name="method_type" invisible="1"/>
+                        <group col="4">
+                            <div class="oe_title" style="width: 390px;" colspan="4">
+                                <label class="oe_edit_only" for="name" string="Product Name"/>
```
Comment (bealdav) [inline external_file_location/task_view.xml]: Why product name ?

File: external_file_location/task.py [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+from .helper import itersubclasses
+from .abstract_task import AbstractTask
+
+
+class Task(models.Model):
+    _name = 'external.file.task'
+    _description = 'Description'
+
+    name = fields.Char()
+    method = fields.Selection(selection='_get_method')
```
Comment (bealdav) [inline external_file_location/task.py]: Needs help parameter in fields

File: external_file_location/helper.py [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+    ... print(cls.__name__)
+    B
+    D
+    E
+    C
+    >>> # get ALL (new-style) classes currently defined
+    >>> [cls.__name__ for cls in itersubclasses(object)] #doctest: +ELLIPSIS
+    ['type', ...'tuple', ...]
+    """
+    #import pdb; pdb.set_trace()
```
Comment (bealdav) [inline external_file_location/helper.py]: pdb here !

File: external_file_location/task.py [hunk 13]
```diff
@@ -0,0 +1,89 @@
...
+        res = []
+        for cls in itersubclasses(AbstractTask):
+            if cls._synchronize_type:
+                cls_info = (cls._key + '_' + cls._synchronize_type,
+                            cls._name + ' ' + cls._synchronize_type)
+                res.append(cls_info)
+        return res
+
+    @api.onchange('method')
+    def onchage_method(self):
```
Comment (bealdav) [inline external_file_location/task.py]: Oups onchage !

File: ir_attachment_metadata/__openerp__.py [hunk 14]
```diff
@@ -0,0 +1,45 @@
...
+{'name': 'ir_attachment_metadata',
+ 'version': '0.0.1',
+ 'author': 'Akretion',
+ 'website': 'www.akretion.com',
+ 'license': 'AGPL-3',
+ 'category': 'Generic Modules',
+ 'description': """
+ Add some useful field to ir.attachment object like:
+ internal and external hash for coherence verification
+ """,
```
Comment (bealdav) [inline ir_attachment_metadata/__openerp__.py]: No more description in **openerp**.py for V8, use module template instead (maintener tools in oca)

File: external_file_location/location_view.xml (hunk 7 above)
Comment (Viggor) [inline external_file_location/location_view.xml:62]: Fixed in **openerp**.py

**Comment by Viggor** [comment]
Need reviews.

**Comment by Viggor** [comment]
This PR is transfered in https://github.com/initOS/connector-interfaces/pull/1 to merge our work and offer a better solution.

**Pull request closed without merging**